### ISSUE TITLE
Test errors (connections failed)

### ISSUE THREAD

**Issue by sagitter** [PythonCharmers/python-future#165]
Hi all,

Some tests are failing due to missing internet connection probably:

`E    ConnectionError: ('Connection aborted.', gaierror(-3, 'Temporary failure in name resolution'))`

They come out during Fedora Koji build-system, not in local.
Log: 

```
=================================== FAILURES ===================================
_________________ TestRequests.test_remove_hooks_then_requests _________________
self = <test_future.test_requests.TestRequests testMethod=test_remove_hooks_then_requests>
    def test_remove_hooks_then_requests(self):
        code = """
                from future import standard_library
                standard_library.install_hooks()

                import builtins
                import http.client
                import html.parser
                """
        with write_module(code, self.tempdir):
            import test_imports_future_stdlib
            standard_library.remove_hooks()
            try:
                import requests
            except ImportError:
                print("Requests doesn't seem to be available. Skipping requests test ...")
            else:
>               r = requests.get('http://google.com')
tests/test_future/test_requests.py:77: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python2.7/site-packages/requests/api.py:69: in get
    return request('get', url, params=params, **kwargs)
/usr/lib/python2.7/site-packages/requests/api.py:50: in request
    response = session.request(method=method, url=url, **kwargs)
/usr/lib/python2.7/site-packages/requests/sessions.py:465: in request
    resp = self.send(prep, **send_kwargs)
/usr/lib/python2.7/site-packages/requests/sessions.py:573: in send
    r = adapter.send(request, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <requests.adapters.HTTPAdapter object at 0xf3f301ac>
request = <PreparedRequest [GET]>, stream = False
timeout = <requests.packages.urllib3.util.timeout.Timeout object at 0xf41238ec>
verify = True, cert = None, proxies = {}
    def send(self, request, stream=False, timeout=None, verify=True, cert=None, proxies=None):
        """Sends PreparedRequest object. Returns Response object.

            :param request: The :class:`PreparedRequest <PreparedRequest>` being sent.
            :param stream: (optional) Whether to stream the request content.
            :param timeout: (optional) How long to wait for the server to send
                data before giving up, as a float, or a (`connect timeout, read
                timeout <user/advanced.html#timeouts>`_) tuple.
            :type timeout: float or tuple
            :param verify: (optional) Whether to verify SSL certificates.
            :param cert: (optional) Any user-provided SSL certificate to be trusted.
            :param proxies: (optional) The proxies dictionary to apply to the request.
            """

        conn = self.get_connection(request.url, proxies)

        self.cert_verify(conn, request.url, verify, cert)
        url = self.request_url(request, proxies)
        self.add_headers(request)

        chunked = not (request.body is None or 'Content-Length' in request.headers)

        if isinstance(timeout, tuple):
            try:
                connect, read = timeout
                timeout = TimeoutSauce(connect=connect, read=read)
            except ValueError as e:
                # this may raise a string formatting error.
                err = ("Invalid timeout {0}. Pass a (connect, read) "
                       "timeout tuple, or a single float to set "
                       "both timeouts to the same value".format(timeout))
                raise ValueError(err)
        else:
            timeout = TimeoutSauce(connect=timeout, read=timeout)

        try:
            if not chunked:
                resp = conn.urlopen(
                    method=request.method,
                    url=url,
                    body=request.body,
                    headers=request.headers,
                    redirect=False,
                    assert_same_host=False,
                    preload_content=False,
                    decode_content=False,
                    retries=self.max_retries,
                    timeout=timeout
                )

            # Send the request.
            else:
                if hasattr(conn, 'proxy_pool'):
                    conn = conn.proxy_pool

                low_conn = conn._get_conn(timeout=timeout)

                try:
                    low_conn.putrequest(request.method,
                                        url,
                                        skip_accept_encoding=True)

                    for header, value in request.headers.items():
                        low_conn.putheader(header, value)

                    low_conn.endheaders()

                    for i in request.body:
                        low_conn.send(hex(len(i))[2:].encode('utf-8'))
                        low_conn.send(b'\r\n')
                        low_conn.send(i)
                        low_conn.send(b'\r\n')
                    low_conn.send(b'0\r\n\r\n')

                    r = low_conn.getresponse()
                    resp = HTTPResponse.from_httplib(
                        r,
                        pool=conn,
                        connection=low_conn,
                        preload_content=False,
                        decode_content=False
                    )
                except:
                    # If we hit any problems here, clean up the connection.
                    # Then, reraise so that we can handle the actual exception.
                    low_conn.close()
                    raise
                else:
                    # All is well, return the connection to the pool.
                    conn._put_conn(low_conn)

        except (ProtocolError, socket.error) as err:
>           raise ConnectionError(err, request=request)
E           ConnectionError: ('Connection aborted.', gaierror(-3, 'Temporary failure in name resolution'))
/usr/lib/python2.7/site-packages/requests/adapters.py:415: ConnectionError
----------------------------- Captured stdout call -----------------------------
Creating /tmp/tmpyTrjFg/test_imports_future_stdlib.py ...
----------------------------- Captured stderr call -----------------------------
requests.packages.urllib3.connectionpool: Starting new HTTP connection (1): google.com
```

Does exist a way to exclude a given test?

Thanks.

**Comment by sagitter** [comment]
Tests failed are

```
test_urllib_request_ssl_redirect
test_urllib_request_http
test_remove_hooks_then_requests
test_requests_cm
```

I'm using this patch to exclude momentarily the above tests:

```
--- tests/test_future/test_standard_library.orig.py	2014-11-21 12:52:03.000000000 +0100
+++ tests/test_future/test_standard_library.py	2015-09-02 11:37:36.808826777 +0200
@@ -318,7 +318,7 @@
         import builtins
         self.assertTrue(hasattr(builtins, 'tuple'))
 
-    # @unittest.skip("ssl support has been stripped out for now ...")
+    @unittest.skip("ConnectionError: ('Connection aborted.', gaierror(-3, 'Temporary failure in name resolution'))...")
     def test_urllib_request_ssl_redirect(self):
         """
         This site redirects to https://...
@@ -332,6 +332,7 @@
         # pprint(r.read().decode('utf-8'))
         self.assertTrue(True)
 
+    @unittest.skip("ConnectionError: ('Connection aborted.', gaierror(-3, 'Temporary failure in name resolution'))...")
     def test_urllib_request_http(self):
         """
         This site (python-future.org) uses plain http (as of 2014-09-23).

--- tests/test_future/test_requests.orig.py	2014-11-21 12:52:03.000000000 +0100
+++ tests/test_future/test_requests.py	2015-09-02 11:39:01.509378296 +0200
@@ -57,6 +57,7 @@
     This class tests whether the requests module conflicts with the
     standard library import hooks, as in issue #19.
     """
+    @unittest.skip("ConnectionError: ('Connection aborted.', gaierror(-3, 'Temporary failure in name resolution'))...")
     def test_remove_hooks_then_requests(self):
         code = """
             from future import standard_library
@@ -79,6 +80,7 @@
             self.assertTrue(True)
 
 
+    @unittest.skip("ConnectionError: ('Connection aborted.', gaierror(-3, 'Temporary failure in name resolution'))...")
     def test_requests_cm(self):
         """
         Tests whether requests can be used importing standard_library modules
```

**Comment by edschofield** [comment]
Hi Antonio, Thanks for your question. I know of ways to skip tests using ``pytest`` and ``nosetests`` but not with the standard Python test runner.

How would the corresponding tests in e.g. Python itself be handled in the Fedora build system?

**Comment by sagitter** [comment]
Hi Ed,

i don't understand what you're asking me, sorry.

If it's useful, this is the full log of ```python-future``` build including test errors: https://kojipkgs.fedoraproject.org//work/tasks/6375/10926375/build.log

**Comment by kislyuk** [comment]
FYI, to conditionally skip tests, you can use unittest skip decorators conditional on things like environment variables: https://docs.python.org/2/library/unittest.html#skipping-tests-and-expected-failures

e.g. something like
```
try:
    r = requests.get('http://google.com')
except Exception:
    online = False

@unittest.skipIf(not online, "Skipping test that relies on a network connection")
def test_foo(...):
    ...
```

**Comment by jmadler** [comment]
Hi!  I believe this may have been resolved in 0.17.  Can you please confirm and reopen if this is still a valid bug?
